### PR TITLE
Feature/cart extra features

### PR DESCRIPTION
**Pull request by namya-jha-17**
cart voucher add /remove
order list
order repeat
order cancel

### CODE REVIEW

File: core/data/lib/src/repository/cart_repository_impl.dart [hunk 1]
```diff
@@ -9,24 +8,52 @@ class CartRepositoryImpl extends CartRepository {
...
+    return await networkPort.removeFromCart(
+        removefromCartRequest: removefromCartRequest);
   }
 
   @override
-  Future<Either<NetworkError, Cart>> removeFromCart({required RemovefromCartRequest removefromCartRequest}) async {
-    return await networkPort.removeFromCart(removefromCartRequest: removefromCartRequest);
+  Future<Either<NetworkError, OrderItem>> checkoutCart(
+      {required CheckoutRequest checkoutRequest}) async {
+    return await networkPort.checkoutCart(checkoutRequest: checkoutRequest);
```
Comment (naikdp7) [inline core/data/lib/src/repository/cart_repository_impl.dart]: avoid unnecessary await

File: core/data/lib/src/repository/cart_repository_impl.dart [hunk 2]
```diff
@@ -9,24 +8,52 @@ class CartRepositoryImpl extends CartRepository {
...
+      {required CartVoucherAddRequest cartVoucherAddRequest}) async {
+    return await networkPort.addVoucherinCart(
+        cartVoucherAddRequest: cartVoucherAddRequest);
+  }
+
+  @override
+  Future<Either<NetworkError, CartVoucherRemove>> removeVoucherfromCart(
+      {required CartVoucherRemoveRequest cartVoucherRemoveRequest}) {
+    // TODO: implement removeVoucherfromCart
+    throw UnimplementedError();
```
Comment (naikdp7) [inline core/data/lib/src/repository/cart_repository_impl.dart]: why its not implemented ?

File: core/domain/lib/src/usecase/cart/cart_address_change_usecase.dart [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+class CartAddressChangeUseCaseParams extends Params {
+  String addressId;
+  CartAddressChangeUseCaseParams(
+    this.addressId,
+  );
+
+  @override
+  Either<AppError, bool> verify() {
+
+    return right(true);
```
Comment (naikdp7) [inline core/domain/lib/src/usecase/cart/cart_address_change_usecase.dart]: validations?

File: core/domain/lib/src/usecase/cart/cart_voucher_add_usecase.dart [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  String cartId;
+  String voucherId;
+  final bool isVoucherIdAvailable;
+
+  CartVoucherUseCaseParams({required this.cartId, 
+    required this.voucherId,required this.isVoucherIdAvailable, });
+
+  @override
+  Either<AppError, bool> verify() {
+    if (Validator.isCartAvailable(this.isVoucherIdAvailable)) {
```
Comment (naikdp7) [inline core/domain/lib/src/usecase/cart/cart_voucher_add_usecase.dart]: why we are sending voucherId in this function?

File: core/domain/lib/src/usecase/cart/cart_voucher_add_usecase.dart [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      return Left(
+        AppError(
+          type: ErrorType.uiEmptyEmail,
+          throwable: Exception(),
+          error: ErrorInfo(message: "voucher is used already"),
```
Comment (naikdp7) [inline core/domain/lib/src/usecase/cart/cart_voucher_add_usecase.dart]: dont provide message from here

File: core/domain/lib/src/usecase/cart/cart_voucher_add_usecase.dart [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+  CartVoucherUseCaseParams({required this.cartId, 
+    required this.voucherId,required this.isVoucherIdAvailable, });
+
+  @override
+  Either<AppError, bool> verify() {
+    if (Validator.isCartAvailable(this.isVoucherIdAvailable)) {
+      return Left(
+        AppError(
+          type: ErrorType.uiEmptyEmail,
```
Comment (naikdp7) [inline core/domain/lib/src/usecase/cart/cart_voucher_add_usecase.dart]: error type is wrong

File: core/domain/lib/src/usecase/orderlist/order_rating_usecase.dart [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  int rating;
+  String review;
+
+
+  OrderRatingUseCaseParam({required this.orderId,
+    required this.rating, required this.review});
+
+  @override
+  Either<AppError, bool> verify() {
+    return Right(true);
```
Comment (naikdp7) [inline core/domain/lib/src/usecase/orderlist/order_rating_usecase.dart]: validations?

File: core/domain/lib/src/usecase/orderlist/order_repeat_usecase.dart [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  }
+
+}
+class OrderRepeatUsecaseParam extends Params {
+  String? orderId;
+  OrderRepeatUsecaseParam({int? orderId});
+
+  @override
+  Either<AppError, bool> verify() {
+    return Right(true);
```
Comment (naikdp7) [inline core/domain/lib/src/usecase/orderlist/order_repeat_usecase.dart]: validations?

File: infrastructure/network-retrofit/lib/src/models/Order_Repeat/order_repeat_response_entity.dart [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+  factory OrderRepeatResponseEntity.fromJson(Map<String, dynamic> json) =>
+      _$OrderRepeatResponseEntityFromJson(json);
+
+  Map<String, dynamic> toJson() => _$OrderRepeatResponseEntityToJson(this);
+
+  @override
+  OrderRepeatResponseEntity restore(OrderRepeat data) {
+   return OrderRepeatResponseEntity(orderRepeatEntity:
+   OrderRepeatEntity(status: data.status));
```
Comment (naikdp7) [inline infrastructure/network-retrofit/lib/src/models/Order_Repeat/order_repeat_response_entity.dart]: please explain this

**Review by naikdp7 — CHANGES_REQUESTED** [review]
Requested changes

File: core/data/lib/src/repository/cart_repository_impl.dart (hunk 1 above)
Comment (namya-jha-17) [reply core/data/lib/src/repository/cart_repository_impl.dart]: ok

File: core/domain/lib/src/usecase/cart/cart_voucher_add_usecase.dart (hunk 4 above)
Comment (namya-jha-17) [reply core/domain/lib/src/usecase/cart/cart_voucher_add_usecase.dart]: it was typo

File: core/domain/lib/src/usecase/cart/cart_voucher_add_usecase.dart (hunk 6 above)
Comment (namya-jha-17) [reply core/domain/lib/src/usecase/cart/cart_voucher_add_usecase.dart]: resolved

File: core/domain/lib/src/usecase/cart/cart_address_change_usecase.dart (hunk 3 above)
Comment (namya-jha-17) [reply core/domain/lib/src/usecase/cart/cart_address_change_usecase.dart]: done

Comment (namya-jha-17) [reply core/domain/lib/src/usecase/cart/cart_address_change_usecase.dart]: done

**Review by naikdp7 — APPROVED** [review]

**Pull request merged**